### PR TITLE
pull client API version negotiation out of the CLI and into the client

### PR DESCRIPTION
**Pull request by rremer**
**- What I did**

Added some tests for client version discovery, and moved that logic out of the CLI and into the client.

The drive here is that I was using the go docker client, and had to call UpdateVersion with a string I needed to get from the server. I noticed that the CLI never had to do this, and sure enough there was some server version negotiation going on in the CLI. At this commit, the CLI just implements a normal client, and the client owns the logic for server version negotiation (so folks consuming the client library can enjoy that same functionality).

**- How I did it**

Added a new conditional statement in client/client.go UpdateClientVersion(str). If the string is empty, it will trigger DowngradeClientVersion() and step into some cut/paste logic that already existed in cli/command/cli.go for pinging the server, grabbing the APIVersion string, and updating that.

**- How to verify it**

```sh
docker run --privileged -e DOCKER_GITCOMMIT=eebfdd0 docker hack/make.sh test-unit test-integration-cli test-docker-py
```

or 
```sh
cd client
go test
```

The manual/functional verification for the CLI built against this commit:
1) start a daemon of an older version
2) docker info

The test cases cover a few more edges than that:
* if the server doesn't return an APIVersion
* if the environment of the CLI specifically pins DOCKER_API_VERSION

**- Description for the changelog**

UpdateClientVersion can auto-downgrade to the server APIVersion

**- A picture of a cute animal (not mandatory but encouraged)**

shibs

### CODE REVIEW

**Comment by vdemeester** [comment]
@rremer build error

```
09:50:33 --- FAIL: TestClientDebugEnabled (0.00s)
09:50:33 	assertions.go:226: 
                          
	Error Trace:	docker_test.go:21
09:50:33 		
	Error:      	Received unexpected error:
09:50:33 		
	            	Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
09:50:33 FAIL
09:50:33 coverage: 29.1% of statements
```

File: client/client.go [hunk 1]
```diff
@@ -218,12 +221,39 @@ func (cli *Client) ClientVersion() string {
...
+// DowngradeClientVersion downgrades the version string associated with this
+// instance of the Client to match the latest version the server supports
+func (cli *Client) DowngradeClientVersion() {
+	ping, _ := cli.Ping(context.Background())
+	cli.DowngradeClientVersionPing(ping)
+}
+
+// DowngradeClientVersionPing downgrades the version string associated with this
+// instance of the Client to match the latest version the server supports
+func (cli *Client) DowngradeClientVersionPing(p types.Ping) {
```
Comment (dnephin) [inline client/client.go]: Same comment about the function name, how about `UpdateAPIVersionFromPing()` ?

The check for `if !cli.manualOverride {...}` needs to move here. It should be the first line in the function I think:

```go
if cli.manualOverride {
    return
}
```

File: client/client_test.go [hunk 2]
```diff
@@ -276,9 +274,117 @@ func TestNewEnvClientSetsDefaultVersion(t *testing.T) {
...
+	// set our version to something new
+	client.UpdateClientVersion("1.25")
+
+	// if no version from server, expect the earliest
+	// version before APIVersion was implemented
+	expected := "1.24"
+
+	// test downgrade
+	client.DowngradeClientVersionPing(ping)
+	if client.version != expected {
```
Comment (dnephin) [inline client/client_test.go]:
```go
assert.Equal(t, expected, client.version)
```

(same with the other assertions)

**Review by dnephin — COMMENTED** [review]
Overall I think this is good, so design LGTM

File: client/client.go [hunk 3]
```diff
@@ -218,12 +221,39 @@ func (cli *Client) ClientVersion() string {
 // UpdateClientVersion updates the version string associated with this
 // instance of the Client. This operation doesn't acquire a mutex.
 func (cli *Client) UpdateClientVersion(v string) {
+	// If no version specified, negotiate with the server
+	if v == "" {
+		cli.DowngradeClientVersion()
+	}
```
Comment (rremer) [inline client/client.go]: Indeed there are no other references to UpdateClientVersion that I can find, however removing it would:
* potentially break existing consumers of the client library
* put update logic into the new negotiate function (although, it is just updating a field)

File: client/client_test.go (hunk 2 above)
Comment (rremer) [inline client/client_test.go]: I was trying to use the same format as the rest of the tests in this suite. If I understood your suggestion, we want to use an assertion framework, so I reworked the tests here to use the same assertion library found in other packages here. I kept this in a separate commit, since it's unrelated cleanup.

**Comment by rremer** [comment]
Thanks for the feedback, @dnephin !

File: client/client.go (hunk 3 above)
Comment (dnephin) [inline client/client.go]: We don't make any promises about not breaking the API on `client/` it's not being released as a versioned package.

File: client/client_test.go (hunk 2 above)
Comment (dnephin) [inline client/client_test.go]: Yes, we are slowly migrating everything to testify assertions.

File: cli/command/cli.go [hunk 4]
```diff
@@ -171,6 +170,7 @@ func (cli *DockerCli) Initialize(opts *cliflags.ClientOptions) error {
 
 			opts.Common.TLSOptions.Passphrase = passwd
 			cli.client, err = NewAPIClientFromFlags(opts.Common, cli.configFile)
+			cli.client.NegotiateAPIVersion()
```
Comment (dnephin) [inline cli/command/cli.go]: This doesn't seem like the right place for this. We also don't want to call ping again. I think this line needs to be removed.

File: cli/command/cli.go [hunk 5]
```diff
@@ -186,23 +186,20 @@ func (cli *DockerCli) Initialize(opts *cliflags.ClientOptions) error {
...
-		// if server version is lower than the current cli, downgrade
-		if versions.LessThan(ping.APIVersion, cli.client.ClientVersion()) {
-			cli.client.UpdateClientVersion(ping.APIVersion)
-		}
+	cli.server = ServerInfo{
+		HasExperimental: ping.Experimental,
+		OSType:          ping.OSType,
 	}
 
+	cli.client.UpdateClientVersion("")
```
Comment (dnephin) [inline cli/command/cli.go]: This should be `NegotiateAPIVersionPing(ping)

File: cli/command/cli.go (hunk 4 above)
Comment (rremer) [inline cli/command/cli.go]: Ah yes, I misunderstood the request to move this to "Initialize". I'm guessing we meant "When we remove UpdateClientVersion, then NegotiateAPIVersion should just take its place".

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
This one needs a rebase because the CLI has been removed from this repo (though the client still exists here).

**Comment by rremer** [comment]
Woah, I leave the country for a bit and the whole project disappears! I've rebased, and once it's merged/released we can update the cli to take advantage of NegotiateAPIVersion() @cpuguy83 

Incidentally, since the project restructure, moby/moby:master has a failing test in client_test.go: TestClientRedirect, happy to investigate that, just wanted to point out that it's unrelated to this change.

File: client/client.go [hunk 6]
```diff
@@ -238,13 +241,29 @@ func (cli *Client) ClientVersion() string {
...
 
-// UpdateClientVersion updates the version string associated with this
-// instance of the Client. This operation doesn't acquire a mutex.
-func (cli *Client) UpdateClientVersion(v string) {
-	if !cli.manualOverride {
-		cli.version = v
+// NegotiateAPIVersion updates the version string associated with this
+// instance of the Client to match the latest version the server supports
+func (cli *Client) NegotiateAPIVersion() {
+	ping, _ := cli.Ping(context.Background())
```
Comment (dnephin) [inline client/client.go]: I guess context should be an argument to this function. the caller can decide if they have a context they'd like to re-use, or if they want to start a new background context.

Comment (rremer) [inline client/client.go]: Done.

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by rremer** [comment]
It appears that node azure-windows-rs1-5 may require daemon/logger/adapter_test.go:137 to be more than 10 milliseconds. Slow shared tenant windows vms aside, we probably don't want to consider log reading to be a failure if it takes more than 10 milliseconds . Maybe >1s is a failure?

**Review by dnephin — APPROVED** [review]
LGTM

And yes, that timeout should be increased. I've seen it fail a few times on windowsRS1

**Comment by rremer** [comment]
FYI, I standardized the timeouts in daemon logger tests in #33727, and implemented the same assertion framework there @dnephin  ... so hopefully you won't have to rebuild as much :p

File: client/client_test.go [hunk 7]
```diff
@@ -189,101 +152,148 @@ func TestParseHost(t *testing.T) {
...
-		os.Setenv(key, envVarValues[key])
+	expected := envMap["DOCKER_API_VERSION"]
+
+	// test that we honored the env var
+	client.NegotiateAPIVersionPing(ping)
+	assert.Equal(t, expected, client.version)
+}
+
+// MapToEnv takes a map of environment variables and sets them
+func MapToEnv(env map[string]string) {
```
Comment (cpuguy83) [inline client/client_test.go]: This should not be exported.

File: client/client_test.go [hunk 8]
```diff
@@ -189,101 +152,148 @@ func TestParseHost(t *testing.T) {
...
+	for k, v := range env {
+		os.Setenv(k, v)
 	}
 }
 
+// EnvToMap returns a map of environment variables
+func EnvToMap() map[string]string {
```
Comment (cpuguy83) [inline client/client_test.go]: This should not be exported.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: client/client_test.go (hunk 8 above)
Comment (rremer) [inline client/client_test.go]: sure, done.

File: client/client.go [hunk 9]
```diff
@@ -238,13 +241,29 @@ func (cli *Client) ClientVersion() string {
...
+// NegotiateAPIVersion updates the version string associated with this
+// instance of the Client to match the latest version the server supports
+func (cli *Client) NegotiateAPIVersion(ctx context.Context) {
+	ping, _ := cli.Ping(ctx)
+	cli.NegotiateAPIVersionPing(ping)
+}
+
+// NegotiateAPIVersionPing updates the version string associated with this
+// instance of the Client to match the latest version the server supports
+func (cli *Client) NegotiateAPIVersionPing(p types.Ping) {
```
Comment (cpuguy83) [inline client/client.go]: Does this need to be exposed on the interface?

Comment (dnephin) [inline client/client.go]: I believe it does. This is what we'll call from `docker/cli` because we already have a `Ping` object, which we need to store `HasExperimental` and `OSType`. 

The other one is a "convenience" for other consumers of this library.

**Review by cpuguy83 — APPROVED** [review]
LGTM
